### PR TITLE
Add support for non-quadratic whiteout-shapes (box and rounded-box style whiteout only)

### PR DESCRIPTION
**Pull request by lfm83**
modified:               lily/grob.cc
       modified:               scm/c++.scm
       modified:               scm/define-grob-properties.scm
       modified:               scm/stencil.scm

### CODE REVIEW

**Comment by uliska** [comment]
Note to anyone: This PR should *not* be merged into `master` here. Once there's an understanding that it's basically good to go someone should upload it into the regular LilyPond review cycle.

File: lily/grob.cc [hunk 1]
```diff
@@ -147,6 +147,7 @@ Grob::get_print_stencil () const
       /* A grob has to be visible, otherwise the whiteout property has no effect. */
       /* Calls the scheme procedure stencil-whiteout in scm/stencils.scm */
       if (!transparent && (scm_is_number (get_property ("whiteout"))
+                           || scm_is_pair (get_property ("whiteout"))  /* We actually would need number-pair */
```
Comment (uliska) [inline lily/grob.cc]: Shouldn't it be easily possible to create a "number-pair" within that OR expression?
I wouldn't mind one or two additional lines when it helps avoid errors down the road or at least clarifies the code

File: scm/define-grob-properties.scm [hunk 2]
```diff
@@ -1168,15 +1168,16 @@ one below this grob.")
 ;;; w
 ;;;
      (when ,ly:moment? "Global time step associated with this column.")
-     (whiteout ,boolean-or-number? "If a number or true, the grob is
-printed over a white background to white-out underlying material, if
+     (whiteout ,boolean-or-number-or-number-pair? "If a number, number-pair
+     or true, the grob is printed over a white background to white-out underlying material, if
```
Comment (uliska) [inline scm/define-grob-properties.scm]: Latest when this is going into the LilyPond review you'll be asked to keep lines to 72 characters.

File: scm/define-grob-properties.scm [hunk 3]
```diff
@@ -1168,15 +1168,16 @@ one below this grob.")
...
 the grob is visible.  A number indicates how far the white background
 extends beyond the bounding box of the grob as a multiple of the
-staff-line thickness.  The @code{LyricHyphen} grob uses a special
+staff-line thickness.  A number pair is interpreted as giving
+X- and Y-extent separately. The @code{LyricHyphen} grob uses a special
```
Comment (uliska) [inline scm/define-grob-properties.scm]: ... and you'll be asked to have two spaces between sentences.

File: scm/define-grob-properties.scm [hunk 4]
```diff
@@ -1168,15 +1168,16 @@ one below this grob.")
...
 implementation of whiteout:  A positive number indicates how far the
 white background extends beyond the bounding box in multiples of
 @code{line-thickness}.  The shape of the background is determined by
-@code{whiteout-style}.  Usually @code{#f} by default. ")
+@code{whiteout-style}. Usually @code{#f} by default. ")
```
Comment (uliska) [inline scm/define-grob-properties.scm]: This too will need to re-add the second space.

**Comment by lfm83** [comment]
Thanks for the comments! Indeed I overlooked the existence of is_number_pair in the C++ environment (the interaction between C++ and Scheme is still a bit of a mystery to me). Changed now.

And sorry, I was not aware of the 72 character rule (but discovered the corresponding section in the CG now). But how to handle periods closing a line? Does a line break substitute for the double whitespace?

**Comment by hahnjo** [comment]
This is quite old and doesn't apply cleanly anymore. If it's still relevant, it should be properly submitted via GitLab.

**Pull request closed without merging**